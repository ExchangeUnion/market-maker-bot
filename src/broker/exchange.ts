import {
  ExchangeType,
  OrderType,
  OrderSide,
} from '../enums';
import { Stream } from './stream';
import { BinanceStream } from './binance/stream';
import { OpenDexStream } from './opendex/stream';
import { OpenDexOrder } from './opendex/order';
import { Logger } from '../logger';
import { BinanceAPI } from './binance/api';
import { OpenDexAPI } from './opendex/api';
import { ExchangeAPI, Balance } from './api';
import { BinanceOrder } from './binance/order';

export type OrderRequest = {
  orderId: string,
  baseAsset: string,
  quoteAsset: string,
  orderType: OrderType,
  orderSide: OrderSide,
  quantity: number,
};

export type StopLimitOrderRequest = LimitOrderRequest & {
  stopPrice: number,
};

export type MarketOrderRequest = OrderRequest & {
  price: string,
};

export type LimitOrderRequest = OrderRequest & {
  price: number,
};

class ExchangeBroker {
  private exchange: ExchangeType;
  private priceStreams = new Map<string, Stream>();
  private apiKey: string | undefined;
  private apiSecret: string | undefined;
  private certPath: string | undefined;
  private rpchost: string | undefined;
  private rpcport: number | undefined;
  private logger: Logger;
  private api!: ExchangeAPI;

  constructor(
    {
      exchange,
      apiKey,
      apiSecret,
      certPath,
      logger,
      rpchost,
      rpcport,
    }:
    {
      exchange: ExchangeType,
      logger: Logger,
      apiKey?: string,
      apiSecret?: string,
      certPath?: string,
      rpchost?: string,
      rpcport?: number,
    },
  ) {
    this.exchange = exchange;
    this.apiKey = apiKey;
    this.apiSecret = apiSecret;
    this.certPath = certPath;
    this.logger = logger;
    this.rpchost = rpchost;
    this.rpcport = rpcport;
  }

  public start = async () => {
    this.checkConfig();
    if (this.exchange === ExchangeType.Binance) {
      this.api = new BinanceAPI({
        logger: this.logger,
        apiKey: this.apiKey!,
        apiSecret: this.apiSecret!,
      });
    } else if (this.exchange === ExchangeType.OpenDEX) {
      this.api = new OpenDexAPI({
        logger: this.logger,
        certPath: this.certPath!,
        rpchost: this.rpchost!,
        rpcport: this.rpcport!,
      });
    }
    await this.api.start();
  }

  public getAssets = async (): Promise<Balance[]> => {
    const ownedAssets = await this.api.getAssets();
    return ownedAssets;
  }

  public getPrice = async (
    tradingPair: string,
    callback: (tradingPair: string, price: number) => any,
  ): Promise<void> => {
    if (!this.priceStreams.has(tradingPair)) {
      if (this.exchange === ExchangeType.Binance) {
        const binanceStream = new BinanceStream(this.logger, tradingPair);
        this.priceStreams.set(tradingPair, binanceStream);
        binanceStream.on('price', callback);
        await binanceStream.start();
      } else if (this.exchange === ExchangeType.OpenDEX) {
        const openDexStream = new OpenDexStream();
        this.priceStreams.set(tradingPair, openDexStream);
        openDexStream.on('price', callback);
        await openDexStream.start();
      }
    } else {
      const priceStream = this.priceStreams.get(tradingPair)!;
      priceStream.on('price', callback);
      return Promise.resolve();
    }
  }

  public newOrder = (
    orderRequest: MarketOrderRequest | LimitOrderRequest | StopLimitOrderRequest,
  ): BinanceOrder | OpenDexOrder => {
    return this.api.newOrder(orderRequest);
  }

  public close = async () => {
    const priceStreamClosePromises: Promise<any>[] = [];
    this.priceStreams.forEach((priceStream) => {
      priceStreamClosePromises.push(priceStream.close());
    });
    await Promise.all(priceStreamClosePromises);
    if (this.api) {
      await this.api.stop();
    }
  }

  private checkConfig = () => {
    if (
      this.exchange === ExchangeType.Binance &&
      (!this.apiKey || !this.apiSecret)
    ) {
      throw new Error('Binance exchange type requires you to specify the apiKey and apiSecret');
    } else if (
      this.exchange === ExchangeType.OpenDEX &&
      (!this.certPath || !this.rpchost || !this.rpcport)
    ) {
      throw new Error('OpenDEX exchange type requires you to specify the certPath, rpchost and rpcport');
    }
  }
}

export { ExchangeBroker };
