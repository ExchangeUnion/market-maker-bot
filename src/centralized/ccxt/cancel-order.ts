import { Exchange, Order } from 'ccxt';
import { from, Observable, defer } from 'rxjs';
import { Config } from '../../config';

const cancelOrder$ = (
  exchange: Exchange,
  config: Config,
  orderId: string
): Observable<Order> => {
  return defer(() =>
    from(
      exchange.cancelOrder(
        orderId,
        `${config.CEX_BASEASSET}/${config.CEX_QUOTEASSET}`
      )
    )
  );
};

export { cancelOrder$ };
