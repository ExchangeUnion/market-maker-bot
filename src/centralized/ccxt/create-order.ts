import { Exchange, Order } from 'ccxt';
import { OrderSide } from '../../constants';
import BigNumber from 'bignumber.js';
import { Observable, from, defer } from 'rxjs';
import { Config } from '../../config';

type CreateOrderParams = {
  config: Config;
  exchange: Exchange;
  side: OrderSide;
  quantity: BigNumber;
};

const createOrder$ = ({
  config,
  exchange,
  side,
  quantity,
}: CreateOrderParams): Observable<Order> => {
  return defer(() => {
    const price = undefined;
    const params =
      config.CEX === 'Kraken' ? { trading_agreement: 'agree' } : undefined;
    return from(
      exchange.createMarketOrder(
        `${config.CEX_BASEASSET}/${config.CEX_QUOTEASSET}`,
        side,
        parseFloat(quantity.toFixed(4, BigNumber.ROUND_DOWN)),
        price,
        params
      )
    );
  });
};

export { createOrder$, CreateOrderParams };
