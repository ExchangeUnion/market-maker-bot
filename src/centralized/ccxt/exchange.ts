import ccxt, { Exchange } from 'ccxt';
import { Config } from '../../config';

const getExchange = (config: Config): Exchange => {
  switch (config.CEX) {
    case 'BINANCE':
      return new ccxt.binance({
        apiKey: config.CEX_API_KEY,
        secret: config.CEX_API_SECRET,
      });
    case 'KRAKEN':
      return new ccxt.kraken({
        apiKey: config.CEX_API_KEY,
        secret: config.CEX_API_SECRET,
      });
    default:
      throw new Error(
        'Could not get centralized exchange. Invalide CEX configuration option'
      );
  }
};

export { getExchange };
