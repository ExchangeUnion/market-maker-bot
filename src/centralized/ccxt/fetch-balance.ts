import { Balances, Exchange } from 'ccxt';
import { defer, from, Observable } from 'rxjs';

const fetchBalance$ = (exchange: Exchange): Observable<Balances> => {
  return defer(() => from(exchange.fetchBalance()));
};

export { fetchBalance$ };
