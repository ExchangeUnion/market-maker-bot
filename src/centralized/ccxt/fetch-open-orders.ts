import { Exchange, Order } from 'ccxt';
import { defer, from, Observable } from 'rxjs';
import { Config } from '../../config';

const fetchOpenOrders$ = (
  exchange: Exchange,
  config: Config
): Observable<Order[]> => {
  return defer(() =>
    from(
      exchange.fetchOpenOrders(
        `${config.CEX_BASEASSET}/${config.CEX_QUOTEASSET}`
      )
    )
  );
};

export { fetchOpenOrders$ };
