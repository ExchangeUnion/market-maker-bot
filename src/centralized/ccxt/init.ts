import { Dictionary, Exchange, Market } from 'ccxt';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { Config } from '../../config';

type InitCEXparams = {
  config: Config;
  loadMarkets$: (exchange: Exchange) => Observable<Dictionary<Market>>;
  getExchange: (config: Config) => Exchange;
};

type InitCEXResponse = {
  markets: Dictionary<Market>;
  exchange: Exchange;
};

const initCEX$ = ({
  getExchange,
  config,
  loadMarkets$,
}: InitCEXparams): Observable<InitCEXResponse> => {
  const exchange = getExchange(config);
  return loadMarkets$(exchange).pipe(
    map(markets => {
      return {
        markets,
        exchange,
      };
    })
  );
};

export { initCEX$, InitCEXparams, InitCEXResponse };
