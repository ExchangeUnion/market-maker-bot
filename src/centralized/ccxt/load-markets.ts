import { Dictionary, Exchange, Market } from 'ccxt';
import { defer, from, Observable } from 'rxjs';

const loadMarkets$ = (exchange: Exchange): Observable<Dictionary<Market>> => {
  return defer(() => from(exchange.loadMarkets()));
};

export { loadMarkets$ };
