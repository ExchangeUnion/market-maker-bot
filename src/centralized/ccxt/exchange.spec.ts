import ccxt from 'ccxt';
import { testConfig } from '../../test-utils';
import { getExchange } from './exchange';

jest.mock('ccxt');

describe('getExchange', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('initializes Binance with API key and secret', () => {
    const config = {
      ...testConfig(),
      ...{ CEX: 'BINANCE' },
    };
    getExchange(config);
    expect(ccxt.binance).toHaveBeenCalledTimes(1);
    expect(ccxt.binance).toHaveBeenCalledWith(
      expect.objectContaining({
        apiKey: config.CEX_API_KEY,
        secret: config.CEX_API_SECRET,
      })
    );
  });

  it('initializes Kraken with API key and secret', () => {
    const config = {
      ...testConfig(),
      ...{ CEX: 'KRAKEN' },
    };
    getExchange(config);
    expect(ccxt.kraken).toHaveBeenCalledTimes(1);
    expect(ccxt.kraken).toHaveBeenCalledWith(
      expect.objectContaining({
        apiKey: config.CEX_API_KEY,
        secret: config.CEX_API_SECRET,
      })
    );
  });

  it('throws for unsupported exchange', () => {
    expect.assertions(1);
    const config = {
      ...testConfig(),
      ...{ CEX: 'Michael' },
    };
    try {
      getExchange(config);
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });
});
