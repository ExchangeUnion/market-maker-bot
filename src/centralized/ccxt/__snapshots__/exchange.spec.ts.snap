// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getExchange throws for unsupported exchange 1`] = `[Error: Could not get centralized exchange. Invalide CEX configuration option]`;
